File: BaseFSharpBot/BasicBot.Tests/BasicBot_Tests.fs
module BasicBot_Tests

open NUnit.Framework
open FsUnit
open BoardBots.Shared
open BasicBot

type FakeBoard()=
    let mutable internalTokens = 
        [|
            [| PlayerToken.None; PlayerToken.None; PlayerToken.None  |]
            [| PlayerToken.None; PlayerToken.None; PlayerToken.None  |]
            [| PlayerToken.None; PlayerToken.None; PlayerToken.None  |]
        |]

    member this.Tokens
        with get() = internalTokens
        and set(value) = internalTokens <- value

    interface IPlayerBoard with
        member this.TokenAt(position) = 
            internalTokens.[position.Column].[position.Row]

let shouldEqual expected (actual:BoardPosition) =
    (actual.Column, actual.Row) |> should equal expected

let shouldNotEqual unexpected (actual:BoardPosition) =
    (actual.Column, actual.Row) |> should not' (equal unexpected)

let shouldBeIn expected (actual:BoardPosition) = 
    expected |> should contain (actual.Column, actual.Row)

let testBot = new BasicBot()

[<Test>]
let ``When board is empty should play in centre`` () =
    let board = new FakeBoard()
    board |> testBot.TakeTurn |> shouldEqual (1,1)

[<Test>]
let ``When centre is taken should not play in centre`` () =
    let board = new FakeBoard()
    board.Tokens.[1].[1] <- PlayerToken.Opponent
    board |> testBot.TakeTurn |> shouldNotEqual (1,1)

[<Test>]
let ``When centre is taken should play in a corner`` () =
    let board = new FakeBoard()
    board.Tokens.[1].[1] <- PlayerToken.Opponent
    board |> testBot.TakeTurn |> shouldBeIn [(0,0);(0,2);(2,0);(2,2)]

File: HAL/BasicBot.Tests/BasicBot_Tests.cs
using NUnit.Framework;
using BoardBots.Shared;

namespace BasicBot.Tests
{
    // you may want to rename this to reflect what you called your bot
    [TestFixture]
    public class BasicBot_Tests
    {
        [Test]
        public void TakeTurn_EmptyBoard_PlaysInColumn0Row0()
        {
            // arrange
            Hal player = new Hal();

            // act
            var result = player.TakeTurn(new FakeBoard());

            // assert
            Assert.That(result.Column, Is.EqualTo(0));
            Assert.That(result.Row, Is.EqualTo(0));
        }

        [Test]
        public void TakeTurn_Column0Row0AlreadyHasPiece_PlaysInColumn0Row0Anyway()
        {
            // arrange
            Hal player = new Hal();
            FakeBoard partiallyFullBoard = new FakeBoard();
            partiallyFullBoard.Tokens[0, 0] = PlayerToken.Opponent;

            // act
            var result = player.TakeTurn(partiallyFullBoard);

            // assert
            Assert.That(result.Column, Is.EqualTo(0));
            Assert.That(result.Row, Is.EqualTo(0));
        }

        // PlayerBoard we can use for testing
        internal class FakeBoard : IPlayerBoard
        {
            internal PlayerToken[,] Tokens = new PlayerToken[3, 3];

            internal FakeBoard()
            {
                InitialiseBoard();
            }

            public PlayerToken TokenAt(BoardPosition position)
            {
                return Tokens[position.Column, position.Row];
            }

            // Set empty fields to PlayerToken.None, which is what the real board will have (they will not be NULL).
            private void InitialiseBoard()
            {
                for (int columns = 0; columns < 3; columns++)
                {
                    for (int rows = 0; rows < 3; rows++)
                    {
                        Tokens[columns, rows] = PlayerToken.None;
                    }
                }
            }
        }
    }
}
